src/com/amplitude/api/AmplitudeClient.java

```diff
@@ -550,53 +593,91 @@ public void logRevenue(String productId, int quantity, double price, String rece
         } catch (JSONException e) {
         }
 
-        logEvent(REVENUE_EVENT, null, apiProperties, System.currentTimeMillis(), false);
+        logEventAsync(REVENUE_EVENT, null, apiProperties, null, getCurrentTimeMillis(), false);
     }
 
-    public void setUserProperties(JSONObject userProperties) {
-        setUserProperties(userProperties, false);
+    // maintain for backwards compatibility
+    public void setUserProperties(final JSONObject userProperties, final boolean replace){
+        setUserProperties(userProperties);
     }
 
-    public void setUserProperties(final JSONObject userProperties, final boolean replace) {
+    public void setUserProperties(final JSONObject userProperties) {
+        if (userProperties == null || userProperties.length() == 0) {
+            return;
+        }
+
         runOnLogThread(new Runnable() {
             @Override
             public void run() {
-                AmplitudeClient instance = AmplitudeClient.this;
-                if (userProperties == null) {
-                    if (replace) {
-                        instance.userProperties = null;
-                    }
-                    return;
-                }
-
-                // Create deep copy to try and prevent ConcurrentModificationException
-                JSONObject copy;
-                try {
-                    copy = new JSONObject(userProperties.toString());
-                } catch (JSONException e) {
-                    Log.e(TAG, e.toString());
-                    return; // could not create copy, cannot merge
-                } // catch (ConcurrentModificationException e) {}
-
-                JSONObject currentUserProperties = instance.userProperties;
-                if (replace || currentUserProperties == null) {
-                    instance.userProperties = copy;
-                    return;
+                // create shallow copy to try and prevent ConcurrentModificationException
+                JSONObject copy = cloneJSONObject(userProperties);
```

Comment:
I noticed logEventAsync was shallow copying the event properties JSON, so I thought it might make sense to do the same for user properties. But now that I think about it, clients are probably more likely to maintain a persistent user property JSON object. Switching back to deep copy.